Using map to store the listener in ConnectionHandlerImpl

soulxu opened the pull request:
Commit Message: Using map to store the listener in ConnectionHandlerImpl
Additional Description:
Risk Level: low
Testing: unittest
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: no

Signed-off-by: He Jie Xu <hejie.xu@intel.com>

soulxu commented on test/server/connection_handler_test.cc:
```diff
@@ -787,7 +803,7 @@ TEST_F(ConnectionHandlerTest, FallbackToWildcardListener) {
   EXPECT_CALL(*access_log_, log(_, _, _, _));
 }
 
-TEST_F(ConnectionHandlerTest, OldBehaviorMatchFirstWildcardListener) {
```
A note for the reviewer, the original test ensure the first listener matched since the original code using the `list`. After using the unorder map, so this behavior changed. I assume the `MatchFirst` isn't part of API contract from the doc.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -83,16 +81,27 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+
+  ASSERT(!listener_map_by_tag_.contains(config.listenerTag()));
+
+  listener_map_by_tag_.emplace(
+      config.listenerTag(),
+      std::make_pair(config.listenSocketFactory().localAddress(), std::move(details)));
+  // This map only store the new listener.
+  listener_map_by_address_.insert_or_assign(config.listenSocketFactory().localAddress()->asString(),
+                                            &listener_map_by_tag_[config.listenerTag()].second);
```
Can `absl::flat_hash_map` guarantee the stability of value address? May we should use `absl::node_hash_map` for `listener_map_by_tag_` or wrapper the details with `unique_ptr`? 🤔

soulxu replied:
Oh, right, absl::flat_hash_map not guarantee that. thanks for point that out! I can use the unique_ptr, or std::reference_wrapper.

soulxu replied:
@rojkov in the end, I use flat_hash_map + unique_ptr, let me know if you still think `node_hash_map` is a better option.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -83,16 +81,27 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
-      ++listener;
+  auto listener_iter = listener_map_by_tag_.find(listener_tag);
+  if (listener_iter != listener_map_by_tag_.end()) {
+    // listener_map_by_address_ may already update to the new listener. Compare it with the one
+    // which find from listener_map_by_tag_, only delete it when it is same listener.
+    if (listener_map_by_address_[listener_iter->second.first->asString()] ==
+        &listener_iter->second.second) {
+      listener_map_by_address_.erase(listener_iter->second.first->asString());
```
IMO, it is not a good idea to use addresses to distinguish between old and new values. Because the memory may be reallocated. 
May be we should store the related listener tag the the `listener_map_by_address_`.

soulxu replied:
got it.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +121,57 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
 void ConnectionHandlerImpl::stopListeners() {
-  for (auto& listener : listeners_) {
-    listener.second.listener_->shutdownListener();
+  for (auto& iter : listener_map_by_address_) {
```
`listener_map_by_tag_` ?

soulxu replied:
My final goal is to support multiple addresses listener. https://github.com/envoyproxy/envoy/issues/11184, 

The `listener_map_by_tag_`  will nest with a vector. For just iterate all the listeners, I just use the listener_map_by_address_ for more convenience, since it is just a flat map.

wbpcode replied:
If we finally complete it, will there are multiple address index for the same listener? Then we may call `shutdownListener()` method of same listener multiple times ?

In fact we can make the `listener_map_by_tag_` more simple just `absl::flat_hash_map<uint64_t, ActiveListenerDetailsPtr> listeners_`.

The reason why the address is saved in the previous `listeners_` list is to support searching by address. Now that we have a `listener_map_by_address_`, it is no longer necessary.

soulxu replied:
> If we finally complete it, will there are multiple address index for the same listener? Then we may call `shutdownListener()` method of same listener multiple times?

The different address will have different listener. so it won't be called multiple times with shutdownListener().

But I think I still should use `listener_map_by_tag_`, since `listener_map_by_address_` only store the newest listener, the drained listener will be overridden at `addListener`. And the ListenerManager will stop the listener first, then add new listener, but I shouldn't depend on how `ListenerManager` invoke `ConnectionhandlerImpl`, that isn't part of API contract.

> 
> In fact we can make the `listener_map_by_tag_` more simple just `absl::flat_hash_map<uint64_t, ActiveListenerDetailsPtr> listeners_`.

Yea, That is much simple, I plan to add tag and address instance field into `ActiveListenerDetail`

> 
> The reason why the address is saved in the previous `listeners_` list is to support searching by address. Now that we have a `listener_map_by_address_`, it is no longer necessary.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +121,57 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
   disable_listeners_ = true;
-  for (auto& listener : listeners_) {
-    listener.second.listener_->pauseListening();
+  for (auto& iter : listener_map_by_address_) {
+    iter.second->listener_->pauseListening();
```
`listener_map_by_tag_` ? the old `listeners_` list includes all the listeners and may be here we should iterate `listener_map_by_tag_` to replace original logic.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +121,57 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
   disable_listeners_ = false;
-  for (auto& listener : listeners_) {
-    listener.second.listener_->resumeListening();
+  for (auto& iter : listener_map_by_address_) {
+    iter.second->listener_->resumeListening();
```
ditto.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +121,57 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
   listener_reject_fraction_ = reject_fraction;
-  for (auto& listener : listeners_) {
-    listener.second.listener_->listener()->setRejectFraction(reject_fraction);
+  for (auto& iter : listener_map_by_address_) {
+    iter.second->listener_->listener()->setRejectFraction(reject_fraction);
```
ditto.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +121,57 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
-
-  if (listener_it != listeners_.end()) {
-    return Network::InternalListenerOptRef(listener_it->second.internalListener().value().get());
+  auto listener_it = listener_map_by_address_.find(address->asString());
```
nit: asStringView()

soulxu replied:
got it

wbpcode commented:
Thanks. It's great and added some comments. cc @rojkov

wbpcode commented:
/assign

soulxu commented:
@wbpcode thanks for the review!

wbpcode commented:
In addition, we may need distinguish the listener of different protocols.  Check https://github.com/envoyproxy/envoy/pull/17414.

soulxu commented:
> In addition, we may need distinguish the listener of different protocols. Check #17414.

thanks for another good point!

soulxu commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +138,65 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
 }
 
 Network::InternalListenerOptRef
 ConnectionHandlerImpl::findByAddress(const Network::Address::InstanceConstSharedPtr& address) {
```
I'm thinking to change this method name as `findInternalListenerByAddress`, since the assertion as below, this method is only used for internal listener.

soulxu commented on source/server/connection_handler_impl.cc:
```diff
@@ -224,58 +242,46 @@ ConnectionHandlerImpl::getBalancedHandlerByTag(uint64_t listener_tag) {
 
 Network::BalancedConnectionHandlerOptRef
 ConnectionHandlerImpl::getBalancedHandlerByAddress(const Network::Address::Instance& address) {
```
also think about change this method name to getBalancedTCPHandlerByAddress, since this method is only used for TCP listener

jmarantz commented:
After @wbpcode approves I'll assign to a maintainer for the next pass. Thanks!

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -39,60 +39,86 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
-          listener.second.internalListener()->get().updateListenerConfig(config);
-          return;
-        }
+      auto listener_it = listener_map_by_tag_.find(overridden_listener.value());
+      if (listener_it != listener_map_by_tag_.end()) {
+        listener_it->second->internalListener()->get().updateListenerConfig(config);
+        return;
```
nit:
```
if(auto iter = listener_map_by_tag_.find(overridden_listener.value()); iter != listener_map_by_tag_.end()) {
  // ...
}
```

wbpcode replied:
Personal preference and feel free to just keep original code.

soulxu replied:
done

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -39,60 +39,86 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
-          listener.second.tcpListener()->get().updateListenerConfig(config);
-          return;
-        }
+      auto listener_it = listener_map_by_tag_.find(overridden_listener.value());
+      if (listener_it != listener_map_by_tag_.end()) {
+        listener_it->second->tcpListener()->get().updateListenerConfig(config);
+        return;
```
ditto.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -39,60 +39,86 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+
+  ASSERT(!listener_map_by_tag_.contains(config.listenerTag()));
+
+  listener_map_by_tag_.emplace(config.listenerTag(), std::move(details));
+  // This map only store the new listener.
+  if (absl::holds_alternative<std::reference_wrapper<ActiveTcpListener>>(
+          listener_map_by_tag_[config.listenerTag()]->typed_listener_)) {
+    tcp_listener_map_by_address_.insert_or_assign(
+        config.listenSocketFactory().localAddress()->asStringView(),
+        listener_map_by_tag_[config.listenerTag()]);
+  } else if (absl::holds_alternative<std::reference_wrapper<ActiveInternalListener>>(
+                 listener_map_by_tag_[config.listenerTag()]->typed_listener_)) {
+    internal_listener_map_by_address_.insert_or_assign(
+        config.listenSocketFactory().localAddress()->asStringView(),
+        listener_map_by_tag_[config.listenerTag()]);
+  }
```
```suggestion
  listener_map_by_tag_.emplace(config.listenerTag(), details);
  // This map only store the new listener.
  if (absl::holds_alternative<std::reference_wrapper<ActiveTcpListener>>(
          details->typed_listener_)) {
    tcp_listener_map_by_address_.insert_or_assign(
        config.listenSocketFactory().localAddress()->asStringView(),
        std::move(details));
  } else if (absl::holds_alternative<std::reference_wrapper<ActiveInternalListener>>(
                 details->typed_listener_)) {
    internal_listener_map_by_address_.insert_or_assign(
        config.listenSocketFactory().localAddress()->asStringView(),
        std::move(details));
  }
```

soulxu replied:
thanks, will update

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -39,60 +39,86 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
-      listener = listeners_.erase(listener);
-    } else {
-      ++listener;
+  auto listener_iter = listener_map_by_tag_.find(listener_tag);
+  if (listener_iter != listener_map_by_tag_.end()) {
+    // listener_map_by_address_ may already update to the new listener. Compare it with the one
+    // which find from listener_map_by_tag_, only delete it when it is same listener.
+    if (tcp_listener_map_by_address_.contains(listener_iter->second->address_->asStringView()) &&
+        tcp_listener_map_by_address_[listener_iter->second->address_->asStringView()]
+                ->listener_tag_ == listener_iter->second->listener_tag_) {
+      tcp_listener_map_by_address_.erase(listener_iter->second->address_->asStringView());
+    } else if (internal_listener_map_by_address_.contains(
+                   listener_iter->second->address_->asStringView()) &&
+               internal_listener_map_by_address_[listener_iter->second->address_->asStringView()]
+                       ->listener_tag_ == listener_iter->second->listener_tag_) {
+      internal_listener_map_by_address_.erase(listener_iter->second->address_->asStringView());
     }
```
```suggestion
  auto listener_iter = listener_map_by_tag_.find(listener_tag);
  if (listener_iter != listener_map_by_tag_.end()) {
    absl::string_view address_view = listener_iter->second->address_->asStringView();
    // listener_map_by_address_ may already update to the new listener. Compare it with the one
    // which find from listener_map_by_tag_, only delete it when it is same listener.
    if (tcp_listener_map_by_address_.contains(address_view) &&
        tcp_listener_map_by_address_[address_view]
                ->listener_tag_ == listener_iter->second->listener_tag_) {
      tcp_listener_map_by_address_.erase(address_view);
    } else if (internal_listener_map_by_address_.contains(
                   address_view) &&
               internal_listener_map_by_address_[address_view]
                       ->listener_tag_ == listener_iter->second->listener_tag_) {
      internal_listener_map_by_address_.erase(address_view);
    }
```

soulxu replied:
thanks, will update too!

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +138,65 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
-  for (auto& listener : listeners_) {
-    listener.second.listener_->shutdownListener();
+  for (auto& iter : listener_map_by_tag_) {
+    if (iter.second->listener_->listener() != nullptr) {
```
when will `listener()` return nullptr? May be keep the original logic is ok?

soulxu replied:
It can be nullptr when you try to stop the listener twice. Although the `shutdownListener` is just going to reset the unique pointer and that is fine, but I think it is better to have a check, seems we can't assume what `shutdownListener` will do in the future.

wbpcode replied:
🤔 SGTM.

soulxu replied:
Seems the ActiveTCPListener have the check of listener is nullptr or not for `pauseListener` https://github.com/envoyproxy/envoy/blob/5f90ef2a615c07f7777342fc24496ff035fc6839/source/server/active_tcp_listener.cc#L113
but the ActiveUDPListener doesn't do that https://github.com/envoyproxy/envoy/blob/5f90ef2a615c07f7777342fc24496ff035fc6839/source/server/active_udp_listener.h#L97

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +138,65 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
   disable_listeners_ = true;
-  for (auto& listener : listeners_) {
-    listener.second.listener_->pauseListening();
+  for (auto& iter : listener_map_by_tag_) {
+    if (iter.second->listener_->listener() != nullptr) {
+      iter.second->listener_->pauseListening();
+    }
```
ditto.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +138,65 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
   disable_listeners_ = false;
-  for (auto& listener : listeners_) {
-    listener.second.listener_->resumeListening();
+  for (auto& iter : listener_map_by_tag_) {
+    if (iter.second->listener_->listener() != nullptr) {
+      iter.second->listener_->resumeListening();
+    }
```
ditto.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +138,65 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
   listener_reject_fraction_ = reject_fraction;
-  for (auto& listener : listeners_) {
-    listener.second.listener_->listener()->setRejectFraction(reject_fraction);
+  for (auto& iter : listener_map_by_tag_) {
+    if (iter.second->listener_->listener() != nullptr) {
+      iter.second->listener_->listener()->setRejectFraction(reject_fraction);
+    }
```
ditto.

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -224,58 +242,46 @@ ConnectionHandlerImpl::getBalancedHandlerByTag(uint64_t listener_tag) {
-                              p.first->ip()->isAnyAddress() &&
-                              p.first->ip()->version() == address.ip()->version();
-                     });
+    std::string addr_str;
+    if (address.ip()->version() == Network::Address::IpVersion::v4) {
+      addr_str = "0.0.0.0:";
+    } else {
+      addr_str = "[::]:";
+    }
+    addr_str += std::to_string(address.ip()->port());
```
May be we can fold this to:

```
const std::string addr_str = address.ip()->version() == Network::Address::IpVersion::v4 ? ... : ... + std::to_string(address.ip()->port());
```

soulxu replied:
will do, thanks

wbpcode reviewed (commented):
This looks good for me overall. And just some minor comments.

soulxu commented:
> This looks good for me overall. And just some minor comments.

thanks for the review!

soulxu commented:
@wbpcode appreciate if you can take a look at again!

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +136,66 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
   disable_listeners_ = true;
-  for (auto& listener : listeners_) {
-    listener.second.listener_->pauseListening();
+  for (auto& iter : listener_map_by_tag_) {
+    if (iter.second->listener_->listener() != nullptr) {
+      iter.second->listener_->pauseListening();
+    }
```
nit: looks like only these codes changed the original logic. can we just add some simple cases to cover it? (It doesn't look like this code affects security, so this is just a small suggestion. )

soulxu replied:
done, thanks

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -198,15 +220,10 @@ ConnectionHandlerImpl::ActiveListenerDetails::internalListener() {
-        listener.second.listener_->listenerTag() == listener_tag) {
-      return listener.second;
-    }
+  auto iter = listener_map_by_tag_.find(listener_tag);
+  if (iter != listener_map_by_tag_.end()) {
```
nit: `if (auto iter = listener_map_by_tag_.find(listener_tag); iter != listener_map_by_tag_.end()) {}`

wbpcode commented on source/server/connection_handler_impl.cc:
```diff
@@ -224,58 +241,42 @@ ConnectionHandlerImpl::getBalancedHandlerByTag(uint64_t listener_tag) {
-                   });
-
+  auto listener_it = tcp_listener_map_by_address_.find(address.asStringView());
   // If there is exact address match, return the corresponding listener.
-  if (listener_it != listeners_.end()) {
+  if (listener_it != tcp_listener_map_by_address_.end()) {
```
```
  if (auto listener_it = tcp_listener_map_by_address_.find(address->asStringView());
      listener_it != tcp_listener_map_by_address_.end()) {}
```

wbpcode reviewed (approved):
Thanks. LGTM. cc @jmarantz @envoyproxy/envoy-maintainers 

Some nit comments. But they are not worth blocking the review.

ggreenway commented on source/server/connection_handler_impl.cc:
```diff
@@ -39,60 +39,84 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
+                   listener_iter->second->listener_tag_) {
+      internal_listener_map_by_address_.erase(address_view);
     }
+    listener_map_by_tag_.erase(listener_tag);
```
It's more efficient to erase with `listener_iter` instead of by tag, so that we don't have to find the element again.

soulxu replied:
got it

ggreenway commented on source/server/connection_handler_impl.cc:
```diff
@@ -112,68 +136,66 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
-    if (listener.second.listener_->listenerTag() == listener_tag) {
-      listener.second.listener_->shutdownListener();
+  if (auto iter = listener_map_by_tag_.find(listener_tag); iter != listener_map_by_tag_.end()) {
+    if (iter->second->listener_->listener() != nullptr) {
```
How can it be in this map, and have a nullptr listener?

soulxu replied:
We will stop the listener before draining it,  the shutdownListener() https://github.com/envoyproxy/envoy/pull/19362/files#diff-fe538327af200510b78cadc759f2c3af989c822a370d3f8a005b92bc59568aa5R153 will reset the pointer. Then it is going to be nullptr.  That means if we shutdown twice, then the second time you will see a nullptr.

ggreenway commented on source/server/connection_handler_impl.cc:
```diff
@@ -224,58 +240,43 @@ ConnectionHandlerImpl::getBalancedHandlerByTag(uint64_t listener_tag) {
+        listener_it->second->tcpListener().value().get());
   }
 
+  absl::optional<std::reference_wrapper<ConnectionHandlerImpl::ActiveListenerDetails>> details;
```
Can this be `Envoy::OptRef`?

soulxu replied:
yes, it can be

ggreenway commented on test/server/connection_handler_test.cc:
```diff
@@ -526,6 +528,27 @@ TEST_F(ConnectionHandlerTest, DisableListener) {
   handler_->disableListeners();
 }
 
+TEST_F(ConnectionHandlerTest, StopAndDisableStoppedListener) {
```
What would cause Envoy to stop or disable an already stopped listener? Please add a comment for the test explaining.

soulxu replied:
Actually, we don't have such case in the current code(as I known), I just make it works in case we have such in the future. 
In tcp listener, there is check to keep the thing safe 
https://github.com/envoyproxy/envoy/blob/35d18c932a325bf62a67b2d95021a18e0e32d167/source/server/active_tcp_listener.cc#L112-L116

But udp listener hasn't such check
https://github.com/envoyproxy/envoy/blob/35d18c932a325bf62a67b2d95021a18e0e32d167/source/server/active_udp_listener.h#L97

So I add the check at ConnectionhandleImpl.

Or I should using the assertion and cleanup the tcp and udp listener check?

ggreenway replied:
I understand now. I think this is fine as is.

ggreenway commented on test/server/connection_handler_test.cc:
```diff
@@ -733,6 +763,74 @@ TEST_F(ConnectionHandlerTest, NormalRedirect) {
+      .WillRepeatedly(ReturnRef(local_address_));
+  handler_->addListener(absl::nullopt, *test_listener1);
+
+  // Listener2 will be replaced by Listener3
```
period at end of sentence

ggreenway commented on test/server/connection_handler_test.cc:
```diff
@@ -733,6 +763,74 @@ TEST_F(ConnectionHandlerTest, NormalRedirect) {
+  EXPECT_CALL(test_listener3->socket_factory_, localAddress())
+      .WillRepeatedly(ReturnRef(listener3_address));
+
+  // This emulated the case of update listener in-place. stop the old listener and
```
Capitalize `stop`, and period at end of sentence.

ggreenway commented on test/server/connection_handler_test.cc:
```diff
@@ -733,6 +763,74 @@ TEST_F(ConnectionHandlerTest, NormalRedirect) {
+        }
+        return true;
+      }));
+  // This is the address of listener2 and listener3
```
period at end

ggreenway commented on test/server/connection_handler_test.cc:
```diff
@@ -733,6 +763,74 @@ TEST_F(ConnectionHandlerTest, NormalRedirect) {
   EXPECT_CALL(*listener1, onDestroy());
 }
 
+TEST_F(ConnectionHandlerTest, MatchLatestListener) {
```
Please add comment on what config change operations would result in this sequence of listener operations.

soulxu replied:
got it

ggreenway reviewed (changes requested):
This looks really good. A few questions, and some comments to add. Thanks for fixing this scalability bottleneck!

/wait

soulxu commented:
@ggreenway thanks for the review!

I addressed your comments except for https://github.com/envoyproxy/envoy/pull/19362#discussion_r788936941, but let me know if you expect the assertion used there.

ggreenway reviewed (approved):
This looks good. Thanks!

howardjohn commented:
When upgrading our Istio test suites to this version everything started going a little wonky. I am not 100% sure what's happening but it looks like listeners are getting mixed up - calls are going to the wrong listeners as far as I can tell This is happening with bind_to_port=false. We were able to isolate the regression to this commit.

cc @lambdai

mattklein123 commented:
It wouldn't surprise me if with bind_to_port = false we now have some ordering variability or something else. Feel free to revert this PR until we sort that out. cc @soulxu

soulxu commented:
I will do some tests on `bind_to_port = false` case. Then let you know when I figure out what happened.

soulxu commented:
> When upgrading our Istio test suites to this version everything started going a little wonky. I am not 100% sure what's happening but it looks like listeners are getting mixed up - calls are going to the wrong listeners as far as I can tell This is happening with bind_to_port=false. We were able to isolate the regression to this commit.
> 
> cc @lambdai

@howardjohn do you have more failure info, or which testcase are failing. I checked the test, we have few tests on `bind_to_port = false` case 

https://github.com/envoyproxy/envoy/blob/a844bafc3130667f79fc1846a2e4b25831bbb5c8/test/server/connection_handler_test.cc#L959

https://github.com/envoyproxy/envoy/blob/a844bafc3130667f79fc1846a2e4b25831bbb5c8/test/server/connection_handler_test.cc#L771

I only thought the 0 port address can be a problem. But that shouldn't the case, since I can't create multiple 0 port addresses with bind_to_port as false.

howardjohn commented:
Not Envoy tests, but Istio tests. It's pretty much all of our tests (
https://github.com/istio/istio/pull/37004) since it seems all requests go
to random listeners

On Thu, Feb 3, 2022 at 8:44 AM xuhj ***@***.***> wrote:

> When upgrading our Istio test suites to this version everything started
> going a little wonky. I am not 100% sure what's happening but it looks like
> listeners are getting mixed up - calls are going to the wrong listeners as
> far as I can tell This is happening with bind_to_port=false. We were able
> to isolate the regression to this commit.
>
> cc @lambdai <https://github.com/lambdai>
>
> @howardjohn <https://github.com/howardjohn> do you have more failure
> info, or which testcase are failing. I checked the test, we have few tests
> on bind_to_port = false case
>
>
> https://github.com/envoyproxy/envoy/blob/a844bafc3130667f79fc1846a2e4b25831bbb5c8/test/server/connection_handler_test.cc#L959
>
>
> https://github.com/envoyproxy/envoy/blob/a844bafc3130667f79fc1846a2e4b25831bbb5c8/test/server/connection_handler_test.cc#L771
>
> I only thought the 0 port address can be a problem. But that shouldn't the
> case, since I can't create multiple 0 port addresses with bind_to_port as
> false.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/19362#issuecomment-1029182979>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEYGXKPSFOXMUHPNQFVRW3UZKWF3ANCNFSM5K5NLPWQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

soulxu commented:
I could be wrong here

https://github.com/envoyproxy/envoy/blob/a844bafc3130667f79fc1846a2e4b25831bbb5c8/source/server/connection_handler_impl.cc#L245-L249

I should check `listener_it->second->listener() != nullptr`, otherwise a draining listener could get new connection balanced to.
